modules/reindex/src/main/java/org/elasticsearch/reindex/AbstractAsyncBulkByScrollAction.java

```diff
@@ -191,6 +203,17 @@
         }
         sourceBuilder.version(needsSourceDocumentVersions);
         sourceBuilder.seqNoAndPrimaryTerm(needsSourceDocumentSeqNoAndPrimaryTerm);
+
+        /*
+         * Do not open scroll if max docs <= scroll size and not resuming on version conflicts
+         */
+        if (mainRequest.getMaxDocs() != -1
+            && mainRequest.getMaxDocs() <= preparedSearchRequest.source().size()
+            && mainRequest.isAbortOnVersionConflict()) {
+            preparedSearchRequest.scroll((Scroll) null);
+        }
+
+        return mainRequest.setSearchRequest(preparedSearchRequest);
```

Comment:
I think this is equally surprising. The client still sees the `ReindexRequest` changing, in case they reuse it for multiple request, the scroll will be gone. I would prefer to do this such that we keep a local new search request instead, for instance by passing such a modified request in to `buildScrollableResultSource`.

Reply:
I could do this, however I am concerned that this class still holds a reference to request and original search request as a result. This could be surprising in certain situations, for example I am checking if there is no scroll to refreshAndFinish early (this could be done other way).